File: majorProject/src/main/java/com/aocc/framework/implementation/AndroidImage.java
package com.aocc.framework.implementation;

import android.graphics.Bitmap;

import com.aocc.framework.Image;
import com.aocc.framework.Graphics.ImageFormat;

//EXCEPT WHERE NOTED, THE FOLLOWING CODE IS SOURCED FROM THE KILOBOLT ANDROID FRAMEWORK

public class AndroidImage implements Image {
    private final Bitmap bitmap;
    private final ImageFormat format;
    /** 1 for standard assets; 2 when loaded from the 2× folder (logical size = pixels ÷ 2). */
    private final int pixelScale;

    public AndroidImage(Bitmap bitmap, ImageFormat format) {
        this(bitmap, format, 1);
    }

    public AndroidImage(Bitmap bitmap, ImageFormat format, int pixelScale) {
        this.bitmap = bitmap;
        this.format = format;
        this.pixelScale = Math.max(1, pixelScale);
    }

    @Override
    public int getWidth() {
        return bitmap.getWidth() / pixelScale;
    }

    @Override
    public int getHeight() {
        return bitmap.getHeight() / pixelScale;
    }

    public int getPixelScale() {
        return pixelScale;
    }

    @Override
    public ImageFormat getFormat() {
        return format;
    }

    public Bitmap getBitmap() {
        return bitmap;
    }

    @Override
    public void dispose() {
        bitmap.recycle();
    }
}

File: majorProject/src/main/kotlin/com/aocc/framework/implementation/AndroidImage.kt
package com.aocc.framework.implementation

import android.graphics.Bitmap
import com.aocc.framework.Graphics.ImageFormat
import com.aocc.framework.Image

//EXCEPT WHERE NOTED, THE FOLLOWING CODE IS SOURCED FROM THE KILOBOLT ANDROID FRAMEWORK

class AndroidImage : Image {
    val bitmap: Bitmap
    override val format: ImageFormat
    /** 1 for standard assets; 2 when loaded from the 2× folder (logical size = pixels ÷ 2). */
    val pixelScale: Int

    constructor(bitmap: Bitmap, format: ImageFormat) : this(bitmap, format, 1)

    constructor(bitmap: Bitmap, format: ImageFormat, pixelScale: Int) {
        this.bitmap = bitmap
        this.format = format
        this.pixelScale = maxOf(1, pixelScale)
    }

    override val width: Int
        get() = bitmap.width / pixelScale

    override val height: Int
        get() = bitmap.height / pixelScale

    override fun dispose() {
        bitmap.recycle()
    }
}
